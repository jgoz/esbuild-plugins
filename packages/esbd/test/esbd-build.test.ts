import { node } from 'execa';
import fs from 'fs';
import path from 'path';
import { beforeAll, describe, expect, it } from 'vitest';

import type { EsbdConfig } from '../lib';

import type { BuildWithHTMLOutput } from './config/serializer';

interface BuildWithHTMLOptions {
  config: Omit<EsbdConfig, 'absWorkingDir' | 'outdir'>;
  files: { [file: string]: string };
}

const TEST_ROOT = path.join(__dirname, '..', 'test-results', 'build');

async function build(options: BuildWithHTMLOptions): Promise<BuildWithHTMLOutput> {
  const absWorkingDir = await fs.promises.mkdtemp(path.join(TEST_ROOT, 'test-'));
  const absOutDir = path.join(absWorkingDir, 'out');

  await fs.promises.mkdir(absOutDir, { recursive: true });

  const config: EsbdConfig = {
    ...options.config,
    sourcemap: false,
    absWorkingDir,
    logLevel: 'warning',
    outdir: './out',
  };

  const bundleFile = path.join(absWorkingDir, 'bundle.js');
  const writeBundle = fs.promises.writeFile(
    bundleFile,
    `require('../../../lib').configure(${JSON.stringify(config)});`,
  );

  const writeFiles = Object.entries(options.files).map(async ([file, content]) => {
    const absFilePath = path.join(absWorkingDir, file);
    await fs.promises.mkdir(path.dirname(absFilePath), { recursive: true });
    await fs.promises.writeFile(absFilePath, content, { encoding: 'utf-8' });
  });

  await Promise.all([...writeFiles, writeBundle]);

  const proc = node(bundleFile, ['build'], {
    encoding: 'utf8',
    reject: false,
    cwd: absWorkingDir,
    env: { ...process.env, NO_COLOR: '1' },
  });

  const { stderr, stdout } = await proc;

  return { cwd: absWorkingDir, outdir: absOutDir, stdout, stderr };
}

describe('build command', () => {
  beforeAll(async () => {
    await fs.promises.mkdir(TEST_ROOT, { recursive: true });
    return async () => {
      await fs.promises.rm(TEST_ROOT, { recursive: true });
    };
  });

  it('builds a simple entry point', () => {
    return expect(
      build({
        config: {
          external: ['react', 'react-dom'],
          entryPoints: { entry: 'src/entry.tsx' },
          format: 'esm',
        },
        files: {
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('includes referenced CSS from entry point', () => {
    return expect(
      build({
        config: {
          external: ['react', 'react-dom'],
          entryPoints: { entry: 'src/entry.tsx', style: 'styles/entry.css' },
          format: 'esm',
        },
        files: {
          'styles/entry.css': `
            @import "./app.css";
            body { background: red; }
          `,
          'styles/app.css': `
            .app { background: green; }
          `,
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('includes referenced CSS from JS', () => {
    return expect(
      build({
        config: {
          external: ['react', 'react-dom'],
          entryPoints: { entry: 'src/entry.tsx' },
          format: 'esm',
        },
        files: {
          'src/entry.css': `
            @import "./app.css";
            body { background: red; }
          `,
          'src/app.css': `
            .app { background: green; }
          `,
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            import './entry.css';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });

  it('supports automatic react runtime', () => {
    return expect(
      build({
        config: {
          external: ['react', 'react-dom'],
          jsx: 'automatic',
          entryPoints: { entry: 'src/entry.tsx' },
          format: 'esm',
        },
        files: {
          'src/entry.tsx': `
            import ReactDOM from 'react-dom';
            import { App } from './app';
            ReactDOM.render(<App />, document.getElementById('root'));
          `,
          'src/app.tsx': `
            export function App() {
              return <div>Hello world</div>;
            }
          `,
        },
      }),
    ).resolves.toMatchSnapshot();
  });
});
