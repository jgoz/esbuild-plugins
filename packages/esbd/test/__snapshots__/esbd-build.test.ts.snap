// Vitest Snapshot v1

exports[`build command > builds a simple entry point 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished entry with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}

// src/entry.tsx
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

`;

exports[`build command > includes referenced CSS from JS 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished entry with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.css
---------------------------------
/* src/app.css */
.app {
  background: green;
}

/* src/entry.css */
body {
  background: red;
}

---------------------------------
entry.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}

// src/entry.tsx
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

`;

exports[`build command > includes referenced CSS from entry point 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished entry, style with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
function App() {
  return /* @__PURE__ */ React.createElement("div", null, "Hello world");
}

// src/entry.tsx
ReactDOM.render(
  /* @__PURE__ */ React.createElement(App, null),
  document.getElementById("root")
);

---------------------------------
style.css
---------------------------------
/* styles/app.css */
.app {
  background: green;
}

/* styles/entry.css */
body {
  background: red;
}

`;

exports[`build command > supports automatic react runtime 1`] = `
---------------------------------
STDOUT
---------------------------------
✔  Finished entry with 0 error(s) and 0 warning(s) in XX time
---------------------------------
entry.js
---------------------------------
// src/entry.tsx
import ReactDOM from "react-dom";

// src/app.tsx
import { jsx } from "react/jsx-runtime";
function App() {
  return /* @__PURE__ */ jsx("div", {
    children: "Hello world",
  });
}

// src/entry.tsx
import { jsx as jsx2 } from "react/jsx-runtime";
ReactDOM.render(/* @__PURE__ */ jsx2(App, {}), document.getElementById("root"));

`;
